File: 0207-course-schedule/0207-course-schedule.swift
class Solution {
    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
                                                
              var array = Array(repeating:[Int](),count:numCourses)
              var indegrees = Array(repeating:0,count:numCourses)
                                                
            for course in prerequisites {
            var tempo:[Int] = array[course[1]]
            tempo.append(course[0])
            array[course[1]] = tempo
            indegrees[course[0]] += 1
            }
            
            var que = [Int]()
            for index in 0..<numCourses {
                if indegrees[index] == 0 {
                que.insert(index, at: 0)
                  }
                }
                var count = numCourses
                while que.count > 0 {
                let current = que.removeLast()
                count -= 1
                let adj = array[current]
                for course in adj {
                indegrees[course] -= 1
                if indegrees[course] == 0 {
                que.insert(course, at: 0)
                   }
                }
            }
                                                
            if count == 0 {
            return true
            } else {
            return false
            }
        }
}

File: 0207-course-schedule/0207-course-schedule.rs
object Solution {
    def canFinish(numCourses: Int, prerequisites: Array[Array[Int]]): Boolean = {
        val inDegree = Array.fill(numCourses)(0)
        var graph = Map((0 until numCourses).map((_,Array[Int]())):_*)
        val visited = Array.fill(numCourses)(false)
        var queue = scala.collection.immutable.Queue[Int]()

        //init inDegree and graph
        for(i <- prerequisites.indices){
          val dad = prerequisites(i)(1)
          val son = prerequisites(i)(0)
          inDegree(son) += 1
          graph = graph + (dad -> graph(dad).:+(son))
        }

        //enqueue the vertexes which inDegree equals 0
        for(i <- inDegree.indices){
          if(inDegree(i) == 0)
            queue = queue.enqueue(i)
        }

        while(queue.nonEmpty){
          val curVertex = queue.dequeue._1
          visited(curVertex) = true
          queue = queue.dequeue._2

          //possible vertexes of current path
          val nextVertexes = graph(curVertex)
          nextVertexes.foreach(v => {
            //remove the edges that related to currentVertex
            inDegree(v) -= 1
            //path go through
            if(inDegree(v) == 0)
              queue = queue.enqueue(v)
          })
        }

        visited.forall(_ == true)
      }
}